compiler/noirc_evaluator/src/ssa/opt/fill_internal_slices.rs

```diff
@@ -0,0 +1,397 @@
+use std::collections::BTreeMap;
+
+use crate::ssa::{
+    ir::{
+        basic_block::BasicBlockId,
+        function::Function,
+        function_inserter::FunctionInserter,
+        instruction::Instruction,
+        post_order::PostOrder,
+        types::Type,
+        value::{Value, ValueId},
+    },
+    ssa_gen::Ssa,
+};
+
+use acvm::FieldElement;
+use fxhash::FxHashMap as HashMap;
+
+impl Ssa {
+    /// Fill out slice values represented by SSA array values to contain
+    /// dummy data that accounts for dynamic array operations in ACIR gen.
+    /// When working with nested slices in ACIR gen it is impossible to discern the size
+    /// of internal slices. Thus, we should use the max size of internal nested slices for reading from memory.
+    /// However, not increasing the capacity of smaller nested slices can lead to errors where
+    /// we end up reading past the specified dynamic index. If we want to also read information that goes
+    /// past the nested slice, we will have garbage data in its place if the nested structure is transformed to match.
+    pub(crate) fn fill_internal_slices(mut self) -> Ssa {
+        for function in self.functions.values_mut() {
+            let mut context = Context::new(function);
+            context.process_blocks();
+        }
+        self
+    }
+}
+
+struct Context<'f> {
+    post_order: PostOrder,
+    inserter: FunctionInserter<'f>,
+}
+
+impl<'f> Context<'f> {
+    fn new(function: &'f mut Function) -> Self {
+        let post_order = PostOrder::with_function(function);
+        let inserter = FunctionInserter::new(function);
+
+        Context { post_order, inserter }
+    }
+
+    fn process_blocks(&mut self) {
+        let mut block_order = PostOrder::with_function(self.inserter.function).into_vec();
+        block_order.reverse();
+        for block in block_order {
+            self.process_block(block);
+        }
+    }
+
+    fn process_block(&mut self, block: BasicBlockId) {
+        // Fetch SSA values potentially with internal slices
+        let instructions = self.inserter.function.dfg[block].take_instructions();
+        let mut slice_values = Vec::new();
+        let mut slice_sizes = HashMap::default();
+        let mut mapped_slice_values = BTreeMap::new();
+        for instruction in instructions.iter() {
+            match &self.inserter.function.dfg[*instruction] {
+                Instruction::ArrayGet { array, .. } => {
+                    let array_typ = self.inserter.function.dfg.type_of_value(*array);
+                    let array_value = &self.inserter.function.dfg[*array];
+                    if matches!(array_value, Value::Array { .. }) && array_typ.contains_slice_element() {
+                        slice_values.push(*array);
+                        self.compute_slice_sizes(*array, &mut slice_sizes);
+                    }
```

Comment:
Forgot to mention - since it looks like you're trying to find each `Value::Array` by going through get and set instructions, perhaps something like the `Instruction::MakeArray` PR would be helpful for this use case? https://github.com/noir-lang/noir/pull/2494. It is something that I never quite got around to finishing but it could help here since right now to find all arrays you'd need to look in all arrayget, arrayset, call, store, and load instructions to find each array while in the PR you'd only need to look for MakeArray instructions. The mem2reg pass has a similar problem.

Reply:
Hmm good call. I will explore this option. Right now dynamic nested slices is already leading to a lot of updates so if does not serious reduce load I may continue with the pass as is for now and then in a later PR consider adding `Instruction::MakeArray` to improve this pass as well as mem2reg.